specification/flag-evaluation/hooks.md

```diff
@@ -2,4 +2,58 @@
 
 ## Overview
 
-// TODO: this is a stub.
+Hooks are a mechanism which can tie into the lifecycle of flag evaluation. They
+operate similarly to middleware in many web frameworks.
+
+### Hook context
+
+Hook context exists to provide hooks with information about the invocation.
+
+> Hook context **MUST** provide: the flag key, flag type, evaluation context, default value and a list of  executed hooks by stage.
+
+> Hook context **SHOULD** provide: provider, client
+
+> flag key, flag type, default value properties **MUST** be immutable.
+
+> The evaluation context **MUST** be mutable only within the `before` hook.
+
+
+### Hook creation and parameters
+
+> Hooks **MUST** specify at least one stage: `before(HookContext, ImmutableState)`, `after(HookContext, FlagEvaluationDetails, , ImmutableState)`, `error(HookContext, Exception, ImmutableState)`, or `finally(HookContext, ImmutableState)`.
+
+> Condition: If `finally` is a reserved word in the language, `afterAll` should be used.
+
+### Hook registration & ordering
+
+> The API, Client and invocation **MUST** have a method for registering hooks which accepts `flag evaluation options`
+
+> Hooks **MUST** be evaluated in the following order: API before, client before, invocation before, invocation after, client after, API after, invocation finally, client finally, API finally. If errors are present, the order is: invocation error, client error, API error.
+
+> When an error occurs in hook evaluation, further evaluation **MUST** stop and error hook evaluation begins.
+
+> If an error is encountered in the error flow, it **MUST** not be returned to the user.
```

Comment:
Throwing from an error stage is something we said could be a possible exception to the "never throw" rule. If we want this, we can make such an exception in the `flag evaluation` spec. I think it has some utility.

@beeme1mr @agentgonzo @moredip what do you guys think?

Reply:
If we're going to have a rule of "never throw", then I don't think we should start adding exceptions to it.

IMO, a hook is a non-critical part of the evaluation flow (the critical part being the flag evaluation/resolution itself). If you throw an error, it means that the `getXyzValue()` won't return an evaluated flag value to the user (bad) IIUI.

Reply:
Is it important to actually return an error in any of the hooks to the caller? What would they do with such an error? Would it be sufficient to merely log the error in the provider?